File: Assets/Packages/core.ui/Scripts/HighscoreUIControl.lua
local Text = UnityEngine.UI.Text
local I18N = Ballance2.Sys.Language.I18N

---分数统计界面控制
---@class HighscoreUIControl : GameLuaObjectHostClass
---@field TextLevelName Text
HighscoreUIControl = ClassicObject:extend()

function HighscoreUIControl:new() 
end
function HighscoreUIControl:Start() 

  self._CurrentIndex = 1
  self._TextItem = {}
  for i = 1, 10, 1 do
    self._TextItem[i] = {
      textName = self.transform:Find('ItemHighscore'..i..'/TextValue'):GetComponent(Text), ---@type Text
      testScore = self.transform:Find('ItemHighscore'..i..'/TextScoreValue'):GetComponent(Text) ---@type Text
    }
  end

  GameUI.HighscoreUIControl = self
end

---加载关卡分数数据
---@param name string|number
function HighscoreUIControl:LoadLevelData(name) 

  self.LevelNames = Game.HighScoreManager.GetLevelNames()

  local setIndex = true
  if type(name) == 'number' then
    name = self.LevelNames[name]
    setIndex = false
  end

  local data = Game.HighScoreManager.GetData(name)
  if data ~= nil then
    if setIndex then
      self._CurrentIndex = IndexOf(self.LevelNames, name)
    end

    for i = 1, 10, 1 do
      local item = data[i]
      if item ~= nil then
        self._TextItem[i].textName.text = item.name
        self._TextItem[i].testScore.text = tostring(item.score)
      else
        self._TextItem[i].textName.text = ''
        self._TextItem[i].testScore.text = ''
      end
    end    

    self.TextLevelName.text = name
  else
    self._CurrentIndex = 1
    for i = 1, 10, 1 do
      self._TextItem[i].textName.text = ''
      self._TextItem[i].testScore.text = ''
    end  
    self.TextLevelName.text = I18N.Tr('ui.noLevelData')
  end
end
function HighscoreUIControl:Next() 
  if self._CurrentIndex < #self.LevelNames then
    self._CurrentIndex = self._CurrentIndex + 1
  else
    self._CurrentIndex = 1
  end
  self:LoadLevelData(self._CurrentIndex)
end
function HighscoreUIControl:Prev() 
  if self._CurrentIndex > 1 then
    self._CurrentIndex = self._CurrentIndex - 1
  else
    self._CurrentIndex = #self.LevelNames
  end
  self:LoadLevelData(self._CurrentIndex)
end


function CreateClass_HighscoreUIControl() 
  return HighscoreUIControl
end

File: Assets/Game/Menu/Scripts/Control/HighscoreUIControl.cs
using System.Collections.Generic;
using Ballance2.Base;
using Ballance2.Game.GamePlay;
using TMPro;
using UnityEngine;
using UnityEngine.InputSystem;
using UnityEngine.UI;

namespace Ballance2.Menu
{
  public class HighscoreUIControl : GameSingletonBehavior<HighscoreUIControl>  {
    public TMP_Text TextLevelName;
    public RectTransform ContentView;
    public InputAction PrevPage;
    public InputAction NextPage;

    private int _CurrentIndex = 0;
    private List<TextItemData> _TextItem = new List<TextItemData>();
    private List<string> LevelNames;

    private class TextItemData {
      public TMP_Text textName;
      public TMP_Text testScore;
    }

    private void Start() {
      _CurrentIndex = 0;
      _TextItem.Clear();
      PrevPage.performed += (c) => Prev();
      NextPage.performed += (c) => Next();
      for (int i = 1; i <= 10; i++)
      {
         _TextItem.Add(new TextItemData {
          textName = ContentView.Find($"ItemHighscore{i}/TextValue").GetComponent<TMP_Text>(),
          testScore = ContentView.Find($"ItemHighscore{i}/TextScoreValue").GetComponent<TMP_Text>()
        });
      }
    }
    private void OnEnable() {
      PrevPage.Enable();
      NextPage.Enable();
    }
    private void OnDisable() {
      PrevPage.Disable();
      NextPage.Disable();
    }

    public void LoadLevelData(int index) {
      LevelNames = HighscoreManager.Instance.GetLevelNames();
      LoadLevelData(LevelNames[index]);
    }
    public void LoadLevelData(string name) {
      LevelNames = HighscoreManager.Instance.GetLevelNames();

      var data = HighscoreManager.Instance.GetData(name);
      if (data != null) {
        _CurrentIndex = LevelNames.IndexOf(name);

        for (int i = 0; i < 10; i++) {
          var item = data[i];
          if (item != null) {
            _TextItem[i].textName.text = item.name;
            _TextItem[i].testScore.text = item.score.ToString();
          } else {
            _TextItem[i].textName.text = "";
            _TextItem[i].testScore.text = "";
          }
        }    

        TextLevelName.text = name;
      } 
      else 
      {
        _CurrentIndex = 0;
        for (int i = 0; i < 10; i++) {
          _TextItem[i].textName.text = "";
          _TextItem[i].testScore.text = "";
        }  
        TextLevelName.text = $"{name}Fail";
      }
    }

    public void Next() {
      if (_CurrentIndex < LevelNames.Count - 1)
        _CurrentIndex++;
      else
        _CurrentIndex = 0;
      LoadLevelData(_CurrentIndex);
    }
    public void Prev() {
      if( _CurrentIndex > 0)
        _CurrentIndex--;
      else
        _CurrentIndex = LevelNames.Count - 1;
      LoadLevelData(_CurrentIndex);
    }
  }
}